src/main/scala/beam/utils/protocolvis/MessageReader.scala

```diff
@@ -0,0 +1,140 @@
+package beam.utils.protocolvis
+
+import beam.utils.csv.GenericCsvReader
+import com.typesafe.scalalogging.LazyLogging
+
+import java.io.Closeable
+import java.nio.file.{Files, Path}
+import java.util
+import scala.util.Try
+
+/**
+  * @author Dmitry Openkov
+  */
+object MessageReader extends LazyLogging {
+
+  def readData(path: Path): (Iterator[RowData], Closeable) = {
+
+    def readSingleFile(path: Path): (Iterator[RowData], Closeable) = {
+      GenericCsvReader.readAs[RowData](path.toString, rowToRowData, _ => true)
+    }
+
+    def findFiles(dir: Path): Seq[Path] = {
+      val FileNum = """actor_messages_(\d+)""".r.unanchored
+      import scala.jdk.CollectionConverters._
+      val paths = Files.newDirectoryStream(dir, "*.actor_messages_*.csv.gz").iterator().asScala.toSeq
+      val pathWithNum = paths
+        .map(path => path -> path.getFileName.toString)
+        .collect {
+          case (path, FileNum(num)) => path -> num.toInt
+        }
+      val sorted = pathWithNum.sortBy { case (_, num) => num }
+      sorted.map { case (path, _) => path }
+    }
+
+    def readFromDir(path: Path): (Iterator[RowData], Closeable) = {
+      val files: Seq[Path] = findFiles(path)
+      val fileIterSeq: Seq[(Iterator[RowData], Closeable)] = files.map(readSingleFile)
+      val (iterator, closables) = fileIterSeq.foldLeft((Iterator[RowData](), List.empty[Closeable])) {
+        case ((accIt, acc), (it, closable)) => (accIt ++ it, acc :+ closable)
+      }
+      (iterator, () => closables.foreach(closable => Try(closable.close())))
+    }
+
+    val isDirectory = Files.isDirectory(path)
+    if (isDirectory) readFromDir(path) else readSingleFile(path)
+
+  }
+
+  private def escapeNull(str: String) = if (str == null) "" else str
+
+  private def rowToRowData(row: java.util.Map[String, String]): RowData = {
+    def extractActor(record: util.Map[String, String], position: String) = {
+      val parent = escapeNull(record.get(s"${position}_parent"))
+      val name = escapeNull(record.get(s"${position}_name"))
+      Actor(parent, name)
+    }
+
+    def getTick(record: util.Map[String, String]) = {
+      val tickStr = record.get("tick")
+      if (tickStr == null || tickStr == "") -1
+      else if (tickStr.contains('.')) Math.round(tickStr.toDouble).toInt
+      else tickStr.toInt
+    }
+
+    def getTriggerId(record: util.Map[String, String]) = {
+      val triggerIdStr = record.get("triggerId")
+      if (triggerIdStr == null || triggerIdStr == "") -1
```

Comment:
Are these valid - should they be logged as warnings if we are resorting to this fallback? You can also go with `record.getOrDefault("triggerId", "-1")` and then there is the question of invalid strings that are not Long - I'm guessing we would want to blow up in that case?

Reply:
Added throwing an exception in case of wrong input.